File: backend/app/schemas/query.py
from typing import Any

from pydantic import BaseModel, Field


class QueryRequest(BaseModel):
    question: str = Field(min_length=1)


class QueryTokenUsage(BaseModel):
    total_tokens: int
    input_tokens: int
    output_tokens: int
    cached_input_tokens: int


class QueryToolCallInfo(BaseModel):
    attempt_number: int
    name: str | None
    args: dict[str, Any] | None
    result: Any | None = None


class QueryResponse(BaseModel):
    sql_query: str
    summary: str
    columns: list[str]
    rows: list[dict[str, Any]]
    row_count: int
    execution_time_ms: int
    truncated: bool
    token_usage: QueryTokenUsage
    attempt_count: int = 1
    repaired: bool = False
    tool_calls: list[QueryToolCallInfo] = Field(default_factory=list)

File: frontend/src/types/query.ts
/**
 * Query types.
 * Mirrors `backend/app/schemas/query.py`
 * (`QueryTokenUsage` / `QueryToolCallInfo` / `QueryResponse`)
 * backed by the `queries`, `query_attempts` and `tokens` tables.
 */
export type QueryTokenUsage = {
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  cached_input_tokens: number;
};

export type QueryToolCallInfo = {
  attempt_number: number;
  name: string | null;
  args: Record<string, unknown> | null;
  result?: unknown;
};

export type QueryResult = {
  sql_query: string;
  summary: string;
  columns: string[];
  rows: Record<string, unknown>[];
  row_count: number;
  execution_time_ms: number;
  truncated: boolean;
  token_usage: QueryTokenUsage;
  attempt_count: number;
  repaired: boolean;
  tool_calls: QueryToolCallInfo[];
};
